src/libraries/System.Private.CoreLib/src/System/IO/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -0,0 +1,406 @@
+            if (!(handle.IsAsync ?? false))
+            {
+                try
+                {
+                    handle.ThreadPoolBinding = ThreadPoolBoundHandle.BindHandle(handle);
+                }
+                catch (Exception ex)
+                {
+                    // If you passed in a synchronous handle and told us to use
+                    // it asynchronously, throw here.
+                    throw new ArgumentException(SR.Arg_HandleNotAsync, nameof(handle), ex);
+                }
+            }
+        }
+
+        protected override void OnInit()
+        {
+            // This is necessary for async IO using IO Completion ports via our
+            // managed Threadpool API's.  This (theoretically) calls the OS's
+            // BindIoCompletionCallback method, and passes in a stub for the
+            // LPOVERLAPPED_COMPLETION_ROUTINE.  This stub looks at the Overlapped
+            // struct for this request and gets a delegate to a managed callback
+            // from there, which it then calls on a threadpool thread.  (We allocate
+            // our native OVERLAPPED structs 2 pointers too large and store EE state
+            // & GC handles there, one to an IAsyncResult, the other to a delegate.)
+            try
+            {
+                _fileHandle.ThreadPoolBinding = ThreadPoolBoundHandle.BindHandle(_fileHandle);
+            }
+            catch (ArgumentException ex)
+            {
+                throw new IOException(SR.IO_BindHandleFailed, ex);
+            }
+            finally
+            {
+                if (_fileHandle.ThreadPoolBinding == null)
+                {
+                    // We should close the handle so that the handle is not open until SafeFileHandle GC
+                    Debug.Assert(!_exposedHandle, "Are we closing handle that we exposed/not own, how?");
+                    _fileHandle.Dispose();
+                }
+            }
+        }
+
+        // called by BufferedStream
+        internal void OnBufferAllocated(byte[] buffer)
+        {
+            Debug.Assert(buffer != null);
+            Debug.Assert(_preallocatedOverlapped == null);
+
+            _preallocatedOverlapped = new PreAllocatedOverlapped(FileStreamCompletionSource.s_ioCallback, this, buffer);
+        }
+
+        SafeFileHandle IFileStreamCompletionSourceStrategy.FileHandle => _fileHandle;
+
+        FileStreamCompletionSource? IFileStreamCompletionSourceStrategy.CurrentOverlappedOwner => _currentOverlappedOwner;
+
+        FileStreamCompletionSource? IFileStreamCompletionSourceStrategy.CompareExchangeCurrentOverlappedOwner(FileStreamCompletionSource? newSource, FileStreamCompletionSource? existingSource)
+            => Interlocked.CompareExchange(ref _currentOverlappedOwner, newSource, existingSource);
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count)).GetAwaiter().GetResult();
```

Comment:
Something else we don't have an issue for but can investigate later.  Rather than doing the async operation and then blocking waiting on the task, we could look at providing the overlapped I/O with an event handle it can signal when it completes.  That could avoid needing this thread to block waiting for another thread pool thread to do work, and instead this thread would just block waiting for the native file I/O to signal that event.

Reply:
Another idea I have in mind is to detect situation when we have requested for async read using overlapped, but OS decided to perform synchronous read anyway:

https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Private.CoreLib/src/System/IO/LegacyFileStreamStrategy.Windows.cs#L773-L775

We would need to verify if the following statement is still true:

https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Private.CoreLib/src/System/IO/LegacyFileStreamStrategy.Windows.cs#L880-L883